docs/DesignDocs/Enabledebugmalloc.md

```diff
@@ -0,0 +1,229 @@
+    header_t* prev;
+
+    /* The alignment passed to memalign() or zero */
+    uint64_t alignment;
+
+    /* Size of user memory */
+    size_t size;
+
+    /* Return addresses obtained by oe_backtrace() */
+    void* addrs[OE_BACKTRACE_MAX];
+    uint64_t num_addrs;
+
+    /* Option if current object is tracked */
+    bool local_tracking;
+
+    /* Padding to make header a multiple of 16 */
+    uint8_t padding[7];
+
+    /* Contains HEADER_MAGIC2 */
+    uint64_t magic2;
+
+    /* User data */
+    uint8_t data[];
+};
+```
+
+We also add a variable
+```c
+bool oe_use_debug_malloc_tracking = false;
+```
+which will be changed by oe_debug_malloc_tracking_start() and
+oe_debug_malloc_tracking_stop(). While doing the allocation, the objective's
+`local_tracking` is marked as `oe_use_debug_malloc_tracking`, then
+oe_debug_malloc_tracking_report() can get the number of all the not yet freed
+objects.
+
+# Sample
+
+```c
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/debugmalloc.h>
+#include <stdio.h>
+
+// Include the trusted debugmalloc header that is generated
+// during the build. This file is generated by calling the
+// sdk tool oeedger8r against the debugmalloc.edl file.
+#include "debugmalloc_t.h"
+
+// This is the function that the host calls. It prints
+// a message in the enclave before calling back out to
+// the host to print a message from there too.
+void enclave_hello()
+{
+    void* buf1, buf2, buf3;
+    int size1, size2;
+
+    oe_debug_malloc_tracking_start();
+    void* buf1 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+    size1 = oe_debug_malloc_tracking_report();
+
+    oe_debug_malloc_tracking_start();
+    void* buf2 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+
+    oe_debug_malloc_tracking_start();
+    void* buf3 = oe_malloc(4096);
+    oe_debug_malloc_tracking_stop();
+    size2 = oe_debug_malloc_tracking_report();
+
+
+    printf("There are %d, %d un-freed objects reported", size1, size2);
+
+    oe_free(buf1);
+    oe_free(buf2);
+    oe_free(buf3);
+
+    // Call back into the host
+    oe_result_t result = host_hello();
+    if (result != OE_OK)
+    {
+        fprintf(
+            stderr,
+            "Call to host_hello failed: result=%u (%s)\n",
+            result,
+            oe_result_str(result));
+    }
+}
+```
+For the local detection, the expected output result is
+`There are 1, 3 un-freed objects reported.`.
+For the global detection, Since this sample has no memory leak at the point of
+enclave termination, the result is 0. If one of the free is missing, the result
+should be "OE_MEMORY_LEAK".
+
+# Note
+- Since debug malloc is a wrapper of the lower memory allocator, It is fully
+composable with a pluggable OE allocator, and it does not require any change to
+the OE allocators for the new features.
+- To improve users' experience, we add the runtime functions that users can use
+to locate the memory leaks. However, the implementation is not thread-safe.
```

Comment:
This is slightly different from the suggestion I made regarding its interaction with the pluggable allocation, which is only that debug malloc does not require thread-aware allocators to behave differently to work with debug malloc.

When we say that debug malloc itself is not thread-safe, we need to then explain what it means to be unsafe (e.g., does it crash? does it corrupt memory?) In this case, there are two aspects:
- The `start`/`stop`/`report` functions being called on multiple threads. e.g., if I call `start`, `stop` and `report` on different threads at the same time, I would expect the report may be inaccurate, but it shouldn't corrupt memory for example. 
- The underlying allocator tracking allocations separately on independent threads. e.g. if a piece of code that calls `report` after `stop` gets run on a different thread, I expect that it still returns me the same result and the underlying allocation doesn't have to do anything special to make this true.

Reply:
Yes you are right, only local detection has thread safety concern. Please have a look again.